### PR TITLE
[VPlan] Implement VPWidenCastRecipe::computeCost(). (NFCI)

### PR DESCRIPTION
**Pull request by ElvisWang123**
This patch implement `VPWidenCastRecipe::computeCost()` and skip cast recipies in the in-loop reduction.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 1]
```diff
@@ -1429,6 +1429,49 @@ void VPWidenCastRecipe::execute(VPTransformState &State) {
...
+    if (R->isMasked())
+      return TTI::CastContextHint::Masked;
+    return TTI::CastContextHint::Normal;
+  };
+
+  TTI::CastContextHint CCH = TTI::CastContextHint::None;
+  // For Trunc, the context is the only user, which must be a
+  // VPWidenStoreRecipe.
+  if (Opcode == Instruction::Trunc || Opcode == Instruction::FPTrunc) {
+    if (!cast<VPValue>(this)->hasMoreThanOneUniqueUser())
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1457]: does the following below work?
```suggestion
    if (hasMoreThanOneUniqueUser())
```

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 2]
```diff
@@ -1429,6 +1429,49 @@ void VPWidenCastRecipe::execute(VPTransformState &State) {
...
+                                               VPCostContext &Ctx) const {
+  auto *SrcTy = cast<VectorType>(
+      ToVectorTy(Ctx.Types.inferScalarType(getOperand(0)), VF));
+  auto *DestTy = cast<VectorType>(ToVectorTy(getResultType(), VF));
+  // Computes the CastContextHint from a VPWidenMemoryRecipe instruction.
+  auto ComputeCCH = [&](VPWidenMemoryRecipe *R) -> TTI::CastContextHint {
+    assert((isa<VPWidenLoadRecipe>(R) || isa<VPWidenStoreRecipe>(R)) &&
+           "Expected a load or a store!");
+
+    if (VF.isScalar())
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1442]: This seems to be missing handling of VPInterleaveRecipe and VPReplicateRecipe?

And returning normal for live ins (the !Loop->contains() case in the legacy cost model)

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 3]
```diff
@@ -1429,6 +1429,49 @@ void VPWidenCastRecipe::execute(VPTransformState &State) {
...
+    if (!cast<VPValue>(this)->hasMoreThanOneUniqueUser())
+      if (VPWidenMemoryRecipe *Store =
+              dyn_cast<VPWidenMemoryRecipe>(*this->user_begin()))
+        CCH = ComputeCCH(Store);
+  }
+  // For Z/Sext, the context is the operand, which must be a VPWidenLoadRecipe.
+  else if (Opcode == Instruction::ZExt || Opcode == Instruction::SExt ||
+           Opcode == Instruction::FPExt) {
+    if (VPWidenMemoryRecipe *Load = dyn_cast<VPWidenMemoryRecipe>(
+            this->getOperand(0)->getDefiningRecipe()))
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1466]:
```suggestion
            getOperand(0)->getDefiningRecipe()))
```

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 1 above)
Comment (ElvisWang123) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Fixed, thanks.

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 2 above)
Comment (ElvisWang123) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1469]: Thanks for catching that. Handling VPInterleaveRecipe and VPReplicateRecipe in `ComputeCCH`.

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 3 above)
Comment (ElvisWang123) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Remved, thanks.

Comment (ElvisWang123) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Removed, thanks.

**Comment by ElvisWang123** [comment]
Gentle ping!

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 4]
```diff
@@ -7258,12 +7258,29 @@ LoopVectorizationPlanner::precomputeCosts(VPlan &Plan, ElementCount VF,
     const auto &ChainOps = RdxDesc.getReductionOpChain(RedPhi, OrigLoop);
     SetVector<Instruction *> ChainOpsAndOperands(ChainOps.begin(),
                                                  ChainOps.end());
+    auto isZExtOrSExt = [](const unsigned Opcode) -> bool {
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:7261]:
```suggestion
    auto IsZExtOrSExt = [](const unsigned Opcode) -> bool {
```

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 5]
```diff
@@ -1462,6 +1462,56 @@ void VPWidenCastRecipe::execute(VPTransformState &State) {
...
+  if ((Opcode == Instruction::Trunc || Opcode == Instruction::FPTrunc) &&
+      !hasMoreThanOneUniqueUser() && getNumUsers() > 0) {
+    if (auto *StoreRecipe = dyn_cast<VPRecipeBase>(*user_begin()))
+      CCH = ComputeCCH(StoreRecipe);
+  }
+  // For Z/Sext, the context is the operand, which must be a VPWidenLoadRecipe,
+  // a VPInterleaveRecipe, a VPReplicateRecipe or a live-in value.
+  else if (Opcode == Instruction::ZExt || Opcode == Instruction::SExt ||
+           Opcode == Instruction::FPExt) {
+    if (getOperand(0)->isLiveIn())
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1500]: Simpler to store getOperand(0) in a variable, avoiding the repeated calls?

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp [hunk 6]
```diff
@@ -7258,12 +7258,29 @@ LoopVectorizationPlanner::precomputeCosts(VPlan &Plan, ElementCount VF,
...
+      return Opcode == Instruction::ZExt || Opcode == Instruction::SExt;
+    };
     // Also include the operands of instructions in the chain, as the cost-model
     // may mark extends as free.
+    //
+    // For ARM, some of the instruction can folded into the reducion
+    // instruction. So we need to mark all folded instructions free.
+    // For example: We can fold reduce(mul(ext(A), ext(B))) into one
+    // instruction.
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:7270]: Would be good if there would be a generic way to check this and avoid hard-coding here, but I cannot think of a better alternative :(

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 4 above)
Comment (ElvisWang123) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: Fixed, thanks.

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 5 above)
Comment (ElvisWang123) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Fixed, thanks.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp (hunk 6 above)
Comment (ElvisWang123) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:7270]: I am working on the VPlan-based pattern match for ARM MVE reduction instructions.
After that we would be able to remove this hard-coded check in `precomputeCost()`.

Comment (ElvisWang123) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:7270]: @fhahn 
Adding a new VPTransform to fold these ARM MVE instruction patterns into a reduction recipe (or a new recipe) might be an alternative method to remove these hard-coding checks in the cost model.
If the above solution looks good to you, I would like to migrate the VPlan-based pattern match from cost model to the VPlan transforms.

Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:7270]: Sounds like a good start, thanks!

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 7]
```diff
@@ -1462,6 +1462,57 @@ void VPWidenCastRecipe::execute(VPTransformState &State) {
...
+      return TTI::CastContextHint::Reversed;
+    if (WidenMemoryRecipe->isMasked())
+      return TTI::CastContextHint::Masked;
+    return TTI::CastContextHint::Normal;
+  };
+
+  VPValue *Operand = getOperand(0);
+  TTI::CastContextHint CCH = TTI::CastContextHint::None;
+  // For Trunc, the context is the only user, which must be a
+  // VPWidenStoreRecipe, a VPInterleaveRecipe ,or a VPReplicateRecipe.
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1491]:
```suggestion
  // For Trunc/FPTrunc, get the context from the only user.
```

as the mentioned restriction isn't checked here

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 8]
```diff
@@ -1462,6 +1462,57 @@ void VPWidenCastRecipe::execute(VPTransformState &State) {
...
+  if ((Opcode == Instruction::Trunc || Opcode == Instruction::FPTrunc) &&
+      !hasMoreThanOneUniqueUser() && getNumUsers() > 0) {
+    if (auto *StoreRecipe = dyn_cast<VPRecipeBase>(*user_begin()))
+      CCH = ComputeCCH(StoreRecipe);
+  }
+  // For Z/Sext, the context is the operand, which must be a VPWidenLoadRecipe,
+  // a VPInterleaveRecipe, a VPReplicateRecipe or a live-in value.
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1498]:
```suggestion
  // For Z/Sext, get the context from the operand.
```

**Review by fhahn — APPROVED** [review]
LGTM, thanks, with a few remaining suggestions to simplify some comments.

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 7 above)
Comment (ElvisWang123) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Fixed, thanks.

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 8 above)
Comment (ElvisWang123) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Fixed, thanks.

**Review by fhahn — APPROVED** [review]
LGTM, thanks!

**Pull request merged**